### PR TITLE
fix bugs and enhance unit test coverage for all functions in pkg/util/validator.go

### LINKED ISSUE #4504: [BUG]subnet中AllowSubnets大写错误无法被检测&&缺少NatOutgoingPolicyRules中SrcIPs、DstIPs为空字符串的处理逻辑
### Kube-OVN Version

v1.12.22

### Kubernetes Version

v1.25.3

### Operation-system/Kernel Version

"Ubuntu 22.04.4 LTS" 6.8.0-40-generic

### Description

对subnet的AllowSubnets大写检查时，传入参数是subnet.Spec.CIDRBlock，应该改成cidr
Snipaste_2024-09-11_10-51-45
如果matchIPStr是空字符串，ipItems:=strings.Split(matchIPStr, ",") 后ipItems为包含一个空字符的切片，len值为1，无法触发len(ipItems) == 0的逻辑
Snipaste_2024-09-11_11-07-58

### Steps To Reproduce

1.subnet的AllowSubnets中设置一个包含大写字母的地址
2.subnet中的NatOutgoingPolicyRules中SrcIPs、DstIPs设置为空字符串

### Current Behavior

1.subnet中AllowSubnets大写错误无法被检测
2.NatOutgoingPolicyRules中SrcIPs、DstIPs为空字符时无法触发返回错误的逻辑

### Expected Behavior

1.subnet中AllowSubnets的地址存在大写错误时能够在对应的处理逻辑中返回error
2.subnet中NatOutgoingPolicyRules中的SrcIPs、DstIPs为空字符时能够有对应的处理逻辑返回error

### PR DESCRIPTION
**Pull request by dolibali**
# Pull Request

- [ ] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Bug fixes
- Tests

## Which issue(s) this PR fixes

Fixes #4504

### CODE REVIEW

**Comment by oilbeater** [comment]
@dolibali please fix the DCO issue

File: pkg/util/validator.go [hunk 1]
```diff
@@ -205,6 +205,9 @@ func validateNatOutgoingPolicyRules(subnet kubeovnv1.Subnet) error {
 }
 
 func validateNatOutGoingPolicyRuleIPs(matchIPStr string) (string, error) {
+	if matchIPStr == "" || strings.TrimSpace(matchIPStr) == "" {
+		return "", errors.New("MatchIPStr format error")
+	}
```
Comment (zhangzujian) [inline pkg/util/validator.go:210]:
```suggestion
	if matchIPStr = strings.TrimSpace(matchIPStr); matchIPStr == "" {
		return "", errors.New("MatchIPStr format error")
	}
```

File: pkg/util/validator.go [hunk 2]
```diff
@@ -205,6 +205,9 @@ func validateNatOutgoingPolicyRules(subnet kubeovnv1.Subnet) error {
...
 	ipItems := strings.Split(matchIPStr, ",")
 	if len(ipItems) == 0 {
```
Comment (zhangzujian) [inline pkg/util/validator.go:212]: This check should be removed.

File: pkg/util/validator.go [hunk 3]
```diff
@@ -205,6 +205,9 @@ func validateNatOutgoingPolicyRules(subnet kubeovnv1.Subnet) error {
 }
 
 func validateNatOutGoingPolicyRuleIPs(matchIPStr string) (string, error) {
+	if matchIPStr == "" || strings.TrimSpace(matchIPStr) == "" {
+		return "", errors.New("MatchIPStr format error")
```
Comment (bobz965) [inline pkg/util/validator.go:209]: 提示 matchIPStr 不应该为空，当前判断和格式化没有关系

File: pkg/util/validator_test.go [hunk 4]
```diff
@@ -71,7 +71,306 @@ func TestValidateSubnet(t *testing.T) {
...
+				Status: kubeovnv1.SubnetStatus{},
+			},
+			err: "subnet gateway 2001:Db8::1 v6 ip address can not contain upper case",
+		},
+		{
+			name: "CICDblockFormalErr",
+			asubnet: kubeovnv1.Subnet{
+				TypeMeta: metav1.TypeMeta{Kind: "Subnet", APIVersion: "kubeovn.io/v1"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "ut-cicdblock-formal-err",
```
Comment (bobz965) [inline pkg/util/validator_test.go:100]: cicdblock -> cidr-block

File: pkg/util/validator_test.go [hunk 5]
```diff
@@ -1073,6 +1426,111 @@ func TestValidateVpc(t *testing.T) {
 			if (err != nil) != tt.wantErr {
 				t.Errorf("got error = %v, but wantErr %v", err, tt.wantErr)
 			}
+			if tt.errMsg != "" && err != nil && err.Error() != tt.errMsg {
+				t.Errorf("expected error message %q, but got %q", tt.errMsg, err.Error())
+			}
+		})
+	}
+}
+func TestValidateNatOutGoingPolicyRuleIPs(t *testing.T) {
```
Comment (zhangzujian) [inline pkg/util/validator_test.go:1435]:
```suggestion
}

func TestValidateNatOutGoingPolicyRuleIPs(t *testing.T) {
```

File: pkg/util/validator_test.go [hunk 6]
```diff
@@ -71,7 +71,306 @@ func TestValidateSubnet(t *testing.T) {
...
+				Status: kubeovnv1.SubnetStatus{},
+			},
+			err: "subnet ut-invalid-cidr-err cidr 192.168.1.0/invalid is invalid",
+		},
+		{
+			name: "ProtocolInvalidErr",
+			asubnet: kubeovnv1.Subnet{
+				TypeMeta: metav1.TypeMeta{Kind: "Subnet", APIVersion: "kubeovn.io/v1"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "ut-protocal-invalid-err",
```
Comment (zhangzujian) [inline pkg/util/validator_test.go:210]:
```suggestion
					Name: "ut-protocol-invalid-err",
```

File: pkg/util/validator_test.go [hunk 7]
```diff
@@ -71,7 +71,306 @@ func TestValidateSubnet(t *testing.T) {
...
+					ExcludeIps:  []string{"2001:db8::A"},
+					Provider:    "ovn",
+					GatewayType: kubeovnv1.GWDistributedType,
+				},
+				Status: kubeovnv1.SubnetStatus{},
+			},
+			err: "subnet exclude ip 2001:db8::A can not contain upper case",
+		},
+		{
+			name: "CIDRblockUppercaseErr",
```
Comment (bobz965) [inline pkg/util/validator_test.go:140]: 命名要么严格驼峰要么中划线

File: pkg/util/validator_test.go [hunk 8]
```diff
@@ -71,7 +71,306 @@ func TestValidateSubnet(t *testing.T) {
...
+			asubnet: kubeovnv1.Subnet{
+				TypeMeta: metav1.TypeMeta{Kind: "Subnet", APIVersion: "kubeovn.io/v1"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "ut-cidrblock-uppercase-err",
```
Comment (bobz965) [inline pkg/util/validator_test.go:144]: cidrblock -> cidr-block

File: pkg/util/validator_test.go [hunk 9]
```diff
@@ -71,7 +71,306 @@ func TestValidateSubnet(t *testing.T) {
...
+					Namespaces:  nil,
+					CIDRBlock:   "2001:Db8::/32",
+					Gateway:     "2001:db8::1",
+					ExcludeIps:  []string{"2001:db8::a"},
+					Provider:    "ovn",
+					GatewayType: kubeovnv1.GWDistributedType,
+				},
+				Status: kubeovnv1.SubnetStatus{},
+			},
+			err: "subnet cidr block 2001:db8::1 v6 ip address can not contain upper case",
```
Comment (bobz965) [inline pkg/util/validator_test.go:159]: 这个提示不对，应该是block的内容

File: pkg/util/validator_test.go [hunk 10]
```diff
@@ -71,7 +71,306 @@ func TestValidateSubnet(t *testing.T) {
...
+				Status: kubeovnv1.SubnetStatus{},
+			},
+			err: "ipv5 is not a valid protocol type",
+		},
+		{
+			name: "ExternalEgressGatewayUpperCaseErr",
+			asubnet: kubeovnv1.Subnet{
+				TypeMeta: metav1.TypeMeta{Kind: "Subnet", APIVersion: "kubeovn.io/v1"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "ut-externalegressgateway-uppercase-err",
```
Comment (bobz965) [inline pkg/util/validator_test.go:233]: 命名需调整

File: pkg/util/validator_test.go [hunk 11]
```diff
@@ -71,7 +71,306 @@ func TestValidateSubnet(t *testing.T) {
...
+					Protocol:    kubeovnv1.ProtocolIPv6,
+					Namespaces:  nil,
+					CIDRBlock:   "2001:db8::/32",
+					Gateway:     "2001:db8::1",
+					ExcludeIps:  []string{"2001:db8::a"},
+					Provider:    "ovn",
+					GatewayType: kubeovnv1.GWDistributedType,
+					NatOutgoingPolicyRules: []kubeovnv1.NatOutgoingPolicyRule{
+						{
+							Match:  kubeovnv1.NatOutGoingPolicyMatch{SrcIPs: "2001:db8::/32,192.168.0.1/24", DstIPs: "2001:db8::/32"}, // 示例源和目的IP范围
```
Comment (bobz965) [inline pkg/util/validator_test.go:317]: 去除中文注释

File: pkg/util/validator_test.go [hunk 12]
```diff
@@ -71,7 +71,306 @@ func TestValidateSubnet(t *testing.T) {
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "utest-u2ointerconnectionipnotincidrcaseerr",
+				},
+				Spec: kubeovnv1.SubnetSpec{
+					Default:              true,
+					Vpc:                  "ovn-cluster",
+					Protocol:             kubeovnv1.ProtocolIPv6,
+					Namespaces:           nil,
+					CIDRBlock:            "2001:db8::/32",
+					Gateway:              "2001:db8::1", // Intentionally contains an uppercase letter
```
Comment (bobz965) [inline pkg/util/validator_test.go:362]: 注释不符合逻辑

File: pkg/util/validator_test.go [hunk 13]
```diff
@@ -71,7 +71,306 @@ func TestValidateSubnet(t *testing.T) {
...
+				Status: kubeovnv1.SubnetStatus{},
+			},
+			err: "logicalGateway and u2oInterconnection can't be opened at the same time",
+		},
+		{
+			name: "InvalidateNatOutgoingPolicyRulesCaseErr",
+			asubnet: kubeovnv1.Subnet{
+				TypeMeta: metav1.TypeMeta{Kind: "Subnet", APIVersion: "kubeovn.io/v1"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "utest-invalidatenatoutgoingpolicyrulescaseerr",
```
Comment (bobz965) [inline pkg/util/validator_test.go:303]: rename

File: pkg/util/validator_test.go [hunk 14]
```diff
@@ -71,7 +71,306 @@ func TestValidateSubnet(t *testing.T) {
...
+				Status: kubeovnv1.SubnetStatus{},
+			},
+			err: "validate nat policy rules src ips 2001:db8::/32,192.168.0.1/24 failed with err match ips 2001:db8::/32,192.168.0.1/24 protocol is not consistent",
+		},
+		{
+			name: "U2OInterconnectionIPUpperCaseErr",
+			asubnet: kubeovnv1.Subnet{
+				TypeMeta: metav1.TypeMeta{Kind: "Subnet", APIVersion: "kubeovn.io/v1"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "utest-u2ointerconnectionipuppercaseerr",
```
Comment (bobz965) [inline pkg/util/validator_test.go:331]: rename

File: pkg/util/validator_test.go [hunk 15]
```diff
@@ -270,6 +569,30 @@ func TestValidateSubnet(t *testing.T) {
 			},
 			err: "10.18.0/16 in allowSubnets is not a valid address",
 		},
+		{
+			name: "AllowSubnetsUppercaseErr",
+			asubnet: kubeovnv1.Subnet{
+				TypeMeta: metav1.TypeMeta{Kind: "Subnet", APIVersion: "kubeovn.io/v1"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "ut-allowsubnets-uppercase-err",
```
Comment (bobz965) [inline pkg/util/validator_test.go:577]: rename

**Review by zhangzujian — APPROVED** [review]

**Review by bobz965 — APPROVED** [review]

**Pull request merged**